onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -47,26 +72,102 @@
   const auto& operand_indices(model_builder.GetOperandIndices());
   const auto& operand_types(model_builder.GetOperandTypes());
 
-  std::vector<uint32_t> input_indices;
+  InlinedVector<uint32_t> input_indices;
   input_indices.push_back(operand_indices.at(input1));  // input 1
   input_indices.push_back(operand_indices.at(input2));  // input 2
 
   if (add_activation) {
     ADD_SCALAR_OPERAND(model_builder, input_indices, fuse_code);
   }
 
-  ORT_RETURN_IF_ERROR(shaper.Eltwise(input1, input2, output));
-  const OperandType output_operand_type(operand_types.at(input1).type, shaper[output],
+  Shape output_shape;
+  ORT_RETURN_IF_ERROR(GetShapeInfoFromNodeArg(model_builder, output, output_shape));
+  shaper.AddShape(output, output_shape);
+  const OperandType output_operand_type(operand_types.at(input1).type, output_shape,
                                         output_scale, output_zero_point);
   ORT_RETURN_IF_ERROR(model_builder.AddOperation(op_type, input_indices,
                                                  {output}, {output_operand_type}));
   return Status::OK();
 }
 
-static Status AddSqueezeOp(ModelBuilder& model_builder,
-                           const std::string& node_name,
-                           const std::string& input, const std::string& output,
-                           std::vector<int32_t> axes) {
+static Status AddNnapiBatchNormalization(ModelBuilder& model_builder,
+                                         const std::string& input1,
+                                         const std::string& input2,
+                                         const std::string& output1,
+                                         const std::string& input3,
+                                         const std::string& output2,
+                                         int32_t fuse_code,
+                                         float output_scale = 0.0f,
+                                         int32_t output_zero_point = 0) {
+  auto& shaper(model_builder.GetShaper());
+  const auto& operand_indices(model_builder.GetOperandIndices());
+  const auto& operand_types(model_builder.GetOperandTypes());
+
+  // Add Nnapi Mul
+  InlinedVector<uint32_t> input_indices;
+  input_indices.push_back(operand_indices.at(input1));
+  input_indices.push_back(operand_indices.at(input2));
+
+  ADD_SCALAR_OPERAND(model_builder, input_indices, ANEURALNETWORKS_FUSED_NONE);
+
+  auto calculate_eltwise_output_shape = [&](const std::string& input1_name,
+                                            const std::string& input2_name,
+                                            const std::string& output_name) -> Status {
+    const Shape& shape1 = shaper[input1_name];
+    const Shape& shape2 = shaper[input2_name];
+
+    // broadcasting support
+    bool shape1IsBigger = shape1.size() >= shape2.size();
```

Comment:
Broadcasting seems like a fairly generic thing. Do we have an existing helper that does that? If not, can we split this out into a generic helper?